File: 1700-number-of-students-unable-to-eat-lunch/1700-number-of-students-unable-to-eat-lunch.go
func countStudents(students []int, sandwiches []int) int {
    n := len(students)
    pref := [2]int{0,0}

    for i:=0;i<n;i++{
        pref[students[i]]++
    }

    for len(sandwiches) > 0 && pref[sandwiches[0]] > 0 {
        for len(students) > 0 && students[0] == sandwiches[0] {
            pref[students[0]]--
            students = students[1:]
            sandwiches = sandwiches[1:]
        }

        for len(students) > 0 && students[0] != sandwiches[0] && pref[sandwiches[0]] > 0 {
            topStudent := students[0]
            students = students[1:]
            students = append(students, topStudent)
        }

    }

    return len(students)
}

File: 1700-number-of-students-unable-to-eat-lunch/1700-number-of-students-unable-to-eat-lunch.rs
impl Solution {
    pub fn count_students(mut students: Vec<i32>, mut sandwiches: Vec<i32>) -> i32 {
        let mut pref: [usize; 2] = [0, 0];

        for st in students.iter() {
            pref[*st as usize] += 1;
        }

        while students.len() > 0 && pref[sandwiches[0] as usize] > 0 {
            
            while students.len() > 0 && students[0] == sandwiches[0] {
                pref[students[0] as usize] -= 1;
                students.remove(0);
                sandwiches.remove(0);
            }

            while sandwiches.len() > 0 && students[0] != sandwiches[0] && pref[sandwiches[0] as usize] > 0 {
                let front = students.remove(0);
                students.push(front);
            }
        }

        return students.len() as i32;
    }
}